From f0721c4a2ec3a7d3e4d217f2c03276195e51c2df Mon Sep 17 00:00:00 2001
From: Philip Wilk <wiryfuture@gmail.com>
Date: Sat, 11 Dec 2021 00:24:54 +0000
Subject: [PATCH] possible query wire only hotfix

---
 dlls/dnsapi/libresolv.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/dlls/dnsapi/libresolv.c b/dlls/dnsapi/libresolv.c
index 0871348fb22..9fcbe08c116 100644
--- a/dlls/dnsapi/libresolv.c
+++ b/dlls/dnsapi/libresolv.c
@@ -90,8 +90,12 @@ static unsigned long map_options( DWORD options )
         FIXME( "option DNS_QUERY_DONT_RESET_TTL_VALUES not implemented\n" );
     if (options & DNS_QUERY_RESERVED)
         FIXME( "option DNS_QUERY_RESERVED not implemented\n" );
-    if (options & DNS_QUERY_WIRE_ONLY)
+    if (options & DNS_QUERY_WIRE_ONLY) 
+    {
+        ret &= ~RES_RECURSE; /* Prevent recursion for the answer */
+        ret &= RES_AAONLY; /* Only accept authoratative answers */
         FIXME( "option DNS_QUERY_WIRE_ONLY not implemented\n" );
+    }
     if (options & DNS_QUERY_NO_WIRE_QUERY)
         FIXME( "option DNS_QUERY_NO_WIRE_QUERY not implemented\n" );
     if (options & DNS_QUERY_BYPASS_CACHE)
-- 
2.33.1

