From f67e6102f99f6d7b00beb3847d7d74aa871e97a1 Mon Sep 17 00:00:00 2001
From: Philip Wilk <wiryfuture@gmail.com>
Date: Fri, 10 Dec 2021 23:13:06 +0000
Subject: [PATCH 2/2] query_wire_only hotfix

---
 dlls/dnsapi/libresolv.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/dlls/dnsapi/libresolv.c b/dlls/dnsapi/libresolv.c
index 9af2c72d9c6..9c3d2312963 100644
--- a/dlls/dnsapi/libresolv.c
+++ b/dlls/dnsapi/libresolv.c
@@ -90,9 +90,12 @@ static unsigned long map_options( DWORD options )
         FIXME( "option DNS_QUERY_DONT_RESET_TTL_VALUES not implemented\n" );
     if (options & DNS_QUERY_RESERVED)
         FIXME( "option DNS_QUERY_RESERVED not implemented\n" );
-    if (options & DNS_QUERY_WIRE_ONLY)
-        /* Need to add smth here */
+    if (options & DNS_QUERY_WIRE_ONLY) 
+    {
+        ret &= ~RES_RECURSE; /* Prevent recursion for the answer */
+        ret &= RES_AAONLY; /* Only accept authoratative answers */
         FIXME( "option DNS_QUERY_WIRE_ONLY not implemented\n" );
+    }
     if (options & DNS_QUERY_NO_WIRE_QUERY)
         FIXME( "option DNS_QUERY_NO_WIRE_QUERY not implemented\n" );
     if (options & DNS_QUERY_BYPASS_CACHE)
-- 
2.33.1

