From 55b756aa2b91d6b8c25a3a07da1d76869e42ee3a Mon Sep 17 00:00:00 2001
From: Philip Wilk <wiryfuture@gmail.com>
Date: Wed, 8 Dec 2021 00:16:57 +0000
Subject: [PATCH] pretend getaddressinfo is working

---
 dlls/ws2_32/protocol.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/ws2_32/protocol.c b/dlls/ws2_32/protocol.c
index 492400b17a7..3664430937c 100644
--- a/dlls/ws2_32/protocol.c
+++ b/dlls/ws2_32/protocol.c
@@ -418,8 +418,8 @@ int WINAPI GetAddrInfoExW( const WCHAR *name, const WCHAR *servname, DWORD names
         FIXME( "Unsupported cancel handle\n" );
 
     ret = getaddrinfoW( name, servname, (struct addrinfo *)hints, result, overlapped, completion_routine );
-    if (ret) return ret;
-    if (handle) *handle = (HANDLE)0xdeadbeef;
+    /*if (ret) return ret;
+    if (handle) *handle = (HANDLE)0xdeadbeef;*/
     return 0;
 }
 
-- 
2.33.1

