From 1570c33b07f41d60eba5be114288c3cdcadf0072 Mon Sep 17 00:00:00 2001
From: Philip Wilk <wiryfuture@gmail.com>
Date: Fri, 3 Dec 2021 23:26:33 +0000
Subject: [PATCH] pretty sure im just breaking wine at this point

---
 dlls/ntdll/unix/system.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/dlls/ntdll/unix/system.c b/dlls/ntdll/unix/system.c
index 91184b12dba..7237b7eb26f 100644
--- a/dlls/ntdll/unix/system.c
+++ b/dlls/ntdll/unix/system.c
@@ -2545,7 +2545,7 @@ NTSTATUS WINAPI NtQuerySystemInformation( SYSTEM_INFORMATION_CLASS class,
     case SystemPerformanceInformation:  /* 2 */
     {
         SYSTEM_PERFORMANCE_INFORMATION spi;
-        static BOOL fixme_written = FALSE;
+        static BOOL fixme_written = FALSE; /* Here */
 
         get_performance_info( &spi );
         len = sizeof(spi);
@@ -2555,6 +2555,8 @@ NTSTATUS WINAPI NtQuerySystemInformation( SYSTEM_INFORMATION_CLASS class,
             else memcpy( info, &spi, len);
         }
         else ret = STATUS_INFO_LENGTH_MISMATCH;
+        return ret;
+        break;
         if(!fixme_written) {
             FIXME("info_class SYSTEM_PERFORMANCE_INFORMATION\n");
             fixme_written = TRUE;
-- 
2.33.1

