From 24443761facff9d01e71b3a58af779e5b7924336 Mon Sep 17 00:00:00 2001
From: wiryfuture <wiryfuture@gmail.com>
Date: Fri, 3 Dec 2021 10:53:21 +0000
Subject: [PATCH] handle log

---
 dlls/ws2_32/protocol.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/ws2_32/protocol.c b/dlls/ws2_32/protocol.c
index 492400b17a7..2f8e7053cbf 100644
--- a/dlls/ws2_32/protocol.c
+++ b/dlls/ws2_32/protocol.c
@@ -415,7 +415,7 @@ int WINAPI GetAddrInfoExW( const WCHAR *name, const WCHAR *servname, DWORD names
     if (timeout)
         FIXME( "Unsupported timeout\n" );
     if (handle)
-        FIXME( "Unsupported cancel handle\n" );
+        FIXME( "Unsupported cancel handle\n %u", handle );
 
     ret = getaddrinfoW( name, servname, (struct addrinfo *)hints, result, overlapped, completion_routine );
     if (ret) return ret;
-- 
2.33.1

