From 71d97946cb072df9b62ed3e9a50879c892d27a8c Mon Sep 17 00:00:00 2001
From: Philip Wilk <wiryfuture@gmail.com>
Date: Wed, 8 Dec 2021 17:35:58 +0000
Subject: [PATCH 2/2] force use of remote dns

---
 dlls/dnsapi/libresolv.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/dlls/dnsapi/libresolv.c b/dlls/dnsapi/libresolv.c
index 9af2c72d9c6..68acf4df6d9 100644
--- a/dlls/dnsapi/libresolv.c
+++ b/dlls/dnsapi/libresolv.c
@@ -91,7 +91,10 @@ static unsigned long map_options( DWORD options )
     if (options & DNS_QUERY_RESERVED)
         FIXME( "option DNS_QUERY_RESERVED not implemented\n" );
     if (options & DNS_QUERY_WIRE_ONLY)
-        /* Need to add smth here */
+        /* only accept authorative answers, deprecated but should have the side effect of using the authoratative server (over the internet)? */
+        ret &= RES_AAONLY;
+        /* hopefully prevents local dns/cache from being used */
+        ret &= ~RES_DNSRCH; 
         FIXME( "option DNS_QUERY_WIRE_ONLY not implemented\n" );
     if (options & DNS_QUERY_NO_WIRE_QUERY)
         FIXME( "option DNS_QUERY_NO_WIRE_QUERY not implemented\n" );
-- 
2.33.1

