From 5a26cb7b6a7734ca1505aea3dbb0a1898133f07e Mon Sep 17 00:00:00 2001
From: Philip Wilk <wiryfuture@gmail.com>
Date: Tue, 7 Dec 2021 22:54:57 +0000
Subject: [PATCH 3/3] this is wrong, but what if we just ignore this lol

---
 dlls/ws2_32/socket.c | 9 ++-------
 1 file changed, 2 insertions(+), 7 deletions(-)

diff --git a/dlls/ws2_32/socket.c b/dlls/ws2_32/socket.c
index 2db441bee3c..7e3a2901427 100644
--- a/dlls/ws2_32/socket.c
+++ b/dlls/ws2_32/socket.c
@@ -1936,13 +1936,8 @@ static DWORD server_ioctl_sock( SOCKET s, DWORD code, LPVOID in_buff, DWORD in_s
         if (WaitForSingleObject( event, INFINITE ) == WAIT_FAILED)
             return -1;
         status = piosb->u.Status;
-    }
-    if (status == STATUS_NOT_SUPPORTED)
-    {
-        FIXME("Unsupported ioctl %x (device=%x access=%x func=%x method=%x)\n",
-              code, code >> 16, (code >> 14) & 3, (code >> 2) & 0xfff, code & 3);
-    }
-    else if (status == STATUS_SUCCESS)
+    } /* What if we just ignore it lol */
+    else 
         *ret_size = piosb->Information;
 
     return NtStatusToWSAError( status );
-- 
2.33.1

