From 59014d1aee52bcad58761c4f9ff38dd5589c33ac Mon Sep 17 00:00:00 2001
From: Philip Wilk <wiryfuture@gmail.com>
Date: Wed, 8 Dec 2021 18:52:04 +0000
Subject: [PATCH] remove ret values from getaddrinfoexw so handle happens

---
 dlls/ws2_32/protocol.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/dlls/ws2_32/protocol.c b/dlls/ws2_32/protocol.c
index 492400b17a7..8e0c5102e88 100644
--- a/dlls/ws2_32/protocol.c
+++ b/dlls/ws2_32/protocol.c
@@ -403,7 +403,7 @@ int WINAPI GetAddrInfoExW( const WCHAR *name, const WCHAR *servname, DWORD names
                            struct timeval *timeout, OVERLAPPED *overlapped,
                            LPLOOKUPSERVICE_COMPLETION_ROUTINE completion_routine, HANDLE *handle )
 {
-    int ret;
+    /*int ret;*/
 
     TRACE( "name %s, servname %s, namespace %u, namespace_id %s)\n",
            debugstr_w(name), debugstr_w(servname), namespace, debugstr_guid(namespace_id) );
@@ -417,9 +417,9 @@ int WINAPI GetAddrInfoExW( const WCHAR *name, const WCHAR *servname, DWORD names
     if (handle)
         FIXME( "Unsupported cancel handle\n" );
 
-    ret = getaddrinfoW( name, servname, (struct addrinfo *)hints, result, overlapped, completion_routine );
-    if (ret) return ret;
-    if (handle) *handle = (HANDLE)0xdeadbeef;
+    getaddrinfoW( name, servname, (struct addrinfo *)hints, result, overlapped, completion_routine );
+    /*if (ret) return ret;
+    if (handle) *handle = (HANDLE)0xdeadbeef;*/
     return 0;
 }
 
-- 
2.33.1

