From 4115d01e6f94eedfd0176632acf137b0e494ec2c Mon Sep 17 00:00:00 2001
From: Philip Wilk <wiryfuture@gmail.com>
Date: Sat, 18 Dec 2021 13:34:09 +0000
Subject: [PATCH] reverse commit 737fe1f

---
 server/fd.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/server/fd.c b/server/fd.c
index c9a21186722..f4b2ac44eb8 100644
--- a/server/fd.c
+++ b/server/fd.c
@@ -566,6 +566,7 @@ static inline void set_fd_epoll_events( struct fd *fd, int user, int events )
     }
     else if (pollfd[user].fd == -1)
     {
+        if (pollfd[user].events) return;  /* stopped waiting on it, don't restart */
         ctl = EPOLL_CTL_ADD;
     }
     else
@@ -664,6 +665,7 @@ static inline void set_fd_epoll_events( struct fd *fd, int user, int events )
     }
     else if (pollfd[user].fd == -1)
     {
+        if (pollfd[user].events) return;  /* stopped waiting on it, don't restart */
         ev[0].flags |= EV_ADD | ((events & POLLIN) ? EV_ENABLE : EV_DISABLE);
         ev[1].flags |= EV_ADD | ((events & POLLOUT) ? EV_ENABLE : EV_DISABLE);
     }
@@ -772,6 +774,7 @@ static inline void set_fd_epoll_events( struct fd *fd, int user, int events )
     }
     else if (pollfd[user].fd == -1)
     {
+        if (pollfd[user].events) return;  /* stopped waiting on it, don't restart */
         ret = port_associate( port_fd, PORT_SOURCE_FD, fd->unix_fd, events, (void *)user );
     }
     else
@@ -1651,6 +1654,7 @@ void set_fd_events( struct fd *fd, int events )
         pollfd[user].events = POLLERR;
         pollfd[user].revents = 0;
     }
+    else if (pollfd[user].fd != -1 || !pollfd[user].events)
     else
     {
         pollfd[user].fd = fd->unix_fd;
-- 
2.33.1

