From d6c1141336a74c0a908db8259636d57c9321c4e3 Mon Sep 17 00:00:00 2001
From: Philip Wilk <wiryfuture@gmail.com>
Date: Sun, 19 Dec 2021 18:20:33 +0000
Subject: [PATCH] Modify commit 737fe1f

---
 server/fd.c | 28 ++++++++++++++++++++++++----
 1 file changed, 24 insertions(+), 4 deletions(-)

diff --git a/server/fd.c b/server/fd.c
index c9a21186722..f5e67bfd899 100644
--- a/server/fd.c
+++ b/server/fd.c
@@ -566,7 +566,13 @@ static inline void set_fd_epoll_events( struct fd *fd, int user, int events )
     }
     else if (pollfd[user].fd == -1)
     {
-        ctl = EPOLL_CTL_ADD;
+        if (pollfd[user].events) 
+        {   /* stopped waiting on it, don't restart */
+            ctl = EPOLL_CTL_DEL;
+        }  
+        else {
+            ctl = EPOLL_CTL_ADD;
+        }
     }
     else
     {
@@ -664,8 +670,16 @@ static inline void set_fd_epoll_events( struct fd *fd, int user, int events )
     }
     else if (pollfd[user].fd == -1)
     {
-        ev[0].flags |= EV_ADD | ((events & POLLIN) ? EV_ENABLE : EV_DISABLE);
-        ev[1].flags |= EV_ADD | ((events & POLLOUT) ? EV_ENABLE : EV_DISABLE);
+        if (pollfd[user].events) 
+        {  /* stopped waiting on it, don't restart */
+            ev[0].flags |= (events & POLLIN) ? EV_ENABLE : EV_DISABLE;
+            ev[1].flags |= (events & POLLOUT) ? EV_ENABLE : EV_DISABLE;
+        }
+        else
+        {
+            ev[0].flags |= EV_ADD | ((events & POLLIN) ? EV_ENABLE : EV_DISABLE);
+            ev[1].flags |= EV_ADD | ((events & POLLOUT) ? EV_ENABLE : EV_DISABLE);
+        }
     }
     else
     {
@@ -772,7 +786,7 @@ static inline void set_fd_epoll_events( struct fd *fd, int user, int events )
     }
     else if (pollfd[user].fd == -1)
     {
-        ret = port_associate( port_fd, PORT_SOURCE_FD, fd->unix_fd, events, (void *)user );
+        ret = port_associate( port_fd, PORT_SOURCE_FD, fd->unix_fd, events, (void *)user ); 
     }
     else
     {
@@ -1651,6 +1665,12 @@ void set_fd_events( struct fd *fd, int events )
         pollfd[user].events = POLLERR;
         pollfd[user].revents = 0;
     }
+    else if (pollfd[user].fd == -1 && pollfd[user].events)
+    {
+        pollfd[user].fd = -1;
+        pollfd[user].events = POLLERR;
+        pollfd[user].revents = 0;
+    }
     else
     {
         pollfd[user].fd = fd->unix_fd;
-- 
2.33.1

