From d99a11b45ea67408847adadff9a6ca3b9af61d3c Mon Sep 17 00:00:00 2001
From: Philip Wilk <wiryfuture@gmail.com>
Date: Wed, 22 Dec 2021 10:58:36 +0000
Subject: [PATCH] ignore unsupported handle

---
 dlls/ws2_32/protocol.c | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/dlls/ws2_32/protocol.c b/dlls/ws2_32/protocol.c
index 239e4d14cd2..9035e2090cd 100644
--- a/dlls/ws2_32/protocol.c
+++ b/dlls/ws2_32/protocol.c
@@ -499,12 +499,18 @@ int WINAPI GetAddrInfoExW( const WCHAR *name, const WCHAR *servname, DWORD names
     if (timeout)
         FIXME( "Unsupported timeout\n" );
     if (handle)
-        FIXME( "Unsupported cancel handle\n" );
+        FIXME( "Unsupported cancel handle (ignoring)\n" );
 
     ret = getaddrinfoW( name, servname, (struct addrinfo *)hints, result, overlapped, completion_routine );
-    if (ret) return ret;
-    if (handle) *handle = (HANDLE)0xdeadbeef;
-    return 0;
+    if (ret) 
+    {
+        *handle = (HANDLE)0xdeadbeef;
+        return ret;
+    }
+    else 
+    {
+        return 0;
+    }
 }
 
 
-- 
2.33.1

