From b4a5423767218af44da246ff90b1761eed831b3a Mon Sep 17 00:00:00 2001
From: Philip Wilk <wiryfuture@gmail.com>
Date: Tue, 21 Dec 2021 23:27:44 +0000
Subject: [PATCH] reselect-patch

---
 dlls/ws2_32/socket.c |  6 ++++++
 server/fd.c          | 26 ++++++++++++++++++++++----
 2 files changed, 28 insertions(+), 4 deletions(-)

diff --git a/dlls/ws2_32/socket.c b/dlls/ws2_32/socket.c
index bf87e7824e5..6f43f50d0d8 100644
--- a/dlls/ws2_32/socket.c
+++ b/dlls/ws2_32/socket.c
@@ -2524,6 +2524,12 @@ int WINAPI WSAPoll( WSAPOLLFD *fds, ULONG count, int timeout )
         SetLastError(WSAEFAULT);
         return SOCKET_ERROR;
     }
+    if (fds->fd < 0)
+    {
+        SetLastError(WSAEINVAL);
+        fds->revents = 0;
+        return SOCKET_ERROR;
+    }
 
     if (!(sync_event = get_sync_event())) return -1;
 
diff --git a/server/fd.c b/server/fd.c
index a09fc9edfcf..fa1d08aff0d 100644
--- a/server/fd.c
+++ b/server/fd.c
@@ -134,6 +134,8 @@ struct epoll_event
   epoll_data_t data;
 };
 
+static void unmount_fd( struct fd *fd );
+
 static inline int epoll_create( int size )
 {
     return syscall( 254 /*NR_epoll_create*/, size );
@@ -544,11 +546,15 @@ static inline void set_fd_epoll_events( struct fd *fd, int user, int events )
 
     if (events == -1)  /* stop waiting on this fd completely */
     {
-        if (pollfd[user].fd == -1) return;  /* already removed */
+        if (pollfd[user].fd == -1) {
+             pollfd[user].revents = 0;
+             return;  /* already removed */
+        }
         ctl = EPOLL_CTL_DEL;
     }
     else if (pollfd[user].fd == -1)
     {
+        pollfd[user].revents = 0;
         ctl = EPOLL_CTL_ADD;
     }
     else
@@ -641,12 +647,16 @@ static inline void set_fd_epoll_events( struct fd *fd, int user, int events )
 
     if (events == -1)  /* stop waiting on this fd completely */
     {
-        if (pollfd[user].fd == -1) return;  /* already removed */
+        if (pollfd[user].fd == -1) {
+            pollfd[user].revents = 0;
+            return;  /* already removed */
+        }
         ev[0].flags |= EV_DELETE;
         ev[1].flags |= EV_DELETE;
     }
     else if (pollfd[user].fd == -1)
     {
+        pollfd[user].revents = 0;
         ev[0].flags |= EV_ADD | ((events & POLLIN) ? EV_ENABLE : EV_DISABLE);
         ev[1].flags |= EV_ADD | ((events & POLLOUT) ? EV_ENABLE : EV_DISABLE);
     }
@@ -750,11 +760,15 @@ static inline void set_fd_epoll_events( struct fd *fd, int user, int events )
 
     if (events == -1)  /* stop waiting on this fd completely */
     {
-        if (pollfd[user].fd == -1) return;  /* already removed */
+        if (pollfd[user].fd == -1) {
+            pollfd[user].revents = 0;
+            return;  /* already removed */
+        }
         port_dissociate( port_fd, PORT_SOURCE_FD, fd->unix_fd );
     }
     else if (pollfd[user].fd == -1)
     {
+        pollfd[user].revents = 0;
         ret = port_associate( port_fd, PORT_SOURCE_FD, fd->unix_fd, events, (void *)user );
     }
     else
@@ -1634,11 +1648,15 @@ void set_fd_events( struct fd *fd, int events )
         pollfd[user].events = POLLERR;
         pollfd[user].revents = 0;
     }
-    else
+    else if (pollfd[user].fd != -1 || !pollfd[user].events)
     {
         pollfd[user].fd = fd->unix_fd;
         pollfd[user].events = events;
     }
+    else
+    {
+        unmount_fd( fd );
+    }
 }
 
 /* prepare an fd for unmounting its corresponding device */
-- 
2.33.1

