From 70d93149c5f1ed08087ad4f1b03b1f3e4341afa9 Mon Sep 17 00:00:00 2001
From: Philip Wilk <wiryfuture@gmail.com>
Date: Thu, 9 Dec 2021 21:15:25 +0000
Subject: [PATCH 2/2] fix patch?

---
 dlls/dnsapi/libresolv.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/dlls/dnsapi/libresolv.c b/dlls/dnsapi/libresolv.c
index 9af2c72d9c6..6645eaeb6b2 100644
--- a/dlls/dnsapi/libresolv.c
+++ b/dlls/dnsapi/libresolv.c
@@ -91,7 +91,10 @@ static unsigned long map_options( DWORD options )
     if (options & DNS_QUERY_RESERVED)
         FIXME( "option DNS_QUERY_RESERVED not implemented\n" );
     if (options & DNS_QUERY_WIRE_ONLY)
-        /* Need to add smth here */
+        /* BLAST and use every server */
+        ret &= RES_BLAST;
+        /* hopefully prevents local dns/cache from being used */
+        ret &= ~RES_DNSRCH; 
         FIXME( "option DNS_QUERY_WIRE_ONLY not implemented\n" );
     if (options & DNS_QUERY_NO_WIRE_QUERY)
         FIXME( "option DNS_QUERY_NO_WIRE_QUERY not implemented\n" );
-- 
2.33.1

