From cfc14abadae91662e4e307de983f8c321904d779 Mon Sep 17 00:00:00 2001
From: Philip Wilk <wiryfuture@gmail.com>
Date: Wed, 15 Dec 2021 22:35:55 +0000
Subject: [PATCH] bodge old patch?

---
 tools/winebuild/import.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/tools/winebuild/import.c b/tools/winebuild/import.c
index c876d51f8e6..2f38788bfd6 100644
--- a/tools/winebuild/import.c
+++ b/tools/winebuild/import.c
@@ -1260,6 +1260,13 @@ void output_stubs( DLLSPEC *spec )
 
             output( "\tsubl $12,%%esp\n" );
             output_cfi( ".cfi_adjust_cfa_offset 12" );
+            
+            /* League stuff idk */
+            output( "\t.byte 0xb8\n" );                               /* mov eax, SYSCALL */
+            output( "\t.long %d\n", i );
+            output( "\t.byte 0x64,0xff,0x15,0xc0,0x00,0x00,0x00\n" ); /* call dword ptr fs:[0C0h] */
+            output( "\tret $%u\n", odp->type == TYPE_STDCALL ? get_args_size( odp ) : 0 );
+            
             if (UsePIC)
             {
                 output( "\tcall %s\n", asm_name("__wine_spec_get_pc_thunk_eax") );
-- 
2.33.1

